Models/TreeModel.cpp

```diff
@@ -133,9 +142,113 @@ int TreeModel::rowCount(const QModelIndex &parent) const {
   return parentItem->child_size();
 }
 
+Qt::DropActions TreeModel::supportedDropActions() const { return Qt::MoveAction | Qt::CopyAction; }
+
+QStringList TreeModel::mimeTypes() const { return QStringList(treeNodeMime()); }
+
+QMimeData *TreeModel::mimeData(const QModelIndexList &indexes) const {
+  QMimeData *mimeData = new QMimeData();
+  QByteArray data;
+
+  QDataStream stream(&data, QIODevice::WriteOnly);
+  QList<QModelIndex> nodes;
+
+  for (const QModelIndex &index : indexes) {
+    if (!index.isValid() || nodes.contains(index)) continue;
+    nodes << index;
+  }
+
+  // rows are moved starting with the lowest so we can create
+  // unique names in the order of insertion
+  std::sort(nodes.begin(), nodes.end(), std::less<QModelIndex>());
+
+  stream << QCoreApplication::applicationPid();
+  stream << nodes.count();
+  for (const QModelIndex &index : nodes) {
+    TreeNode *node = static_cast<TreeNode *>(index.internalPointer());
+    stream << reinterpret_cast<qlonglong>(node) << index.row();
+  }
+  mimeData->setData(treeNodeMime(), data);
+  return mimeData;
+}
+
+bool TreeModel::dropMimeData(const QMimeData *mimeData, Qt::DropAction action, int row, int /*column*/,
+                             const QModelIndex &parent) {
+  if (action != Qt::MoveAction && action != Qt::CopyAction) return false;
+  // ensure the data is in the format we expect
+  if (!mimeData->hasFormat(treeNodeMime())) return false;
+  QByteArray data = mimeData->data(treeNodeMime());
+  QDataStream stream(&data, QIODevice::ReadOnly);
+
+  qint64 senderPid;
```

Comment:
I don't know what the hell you're asking me. `qint64` is clearly a fixed-width 64 bit integer. The rest of it I just copied from the Qt documentation.

Reply:
I mean, you are reading an int, followed by another int, followed by that many `TreeNode*`-int pairs. What the fuck? Why are you manually serializing/deserealizing? This is why we have (1) structs, and (2) protobuf.

Reply:
Ok, yeah that's why I emphasized that it's pointer based and can't drag and drop tree nodes across processes yet. As I said, I am aware that I can just serialize the treenode into bytes. The reason I didn't yet is because if it does end up in another process, we actually may need to do a file copy for some resources (like the sound data file) which we are not set up to do yet.